### PR TITLE
[KYUUBI #5405] [FLINK] Support Flink 1.18

### PR DESCRIPTION
**Pull request by YesOrNo828**
### _Why are the changes needed?_

Resolve: #5405 Support the Flink 1.18

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### _Was this patch authored or co-authored using generative AI tooling?_

No

### CODE REVIEW

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -53,9 +53,9 @@ jobs:
           - '3.3'
           - '3.4'
           - '3.5'
-        spark-archive: [""]
-        exclude-tags: [""]
-        comment: ["normal"]
+        spark-archive: [ "" ]
+        exclude-tags: [ "" ]
+        comment: [ "normal" ]
```
Comment (pan3793) [inline .github/workflows/master.yml:58]: please avoid such unnecessary changes.

File: pom.xml [hunk 2]
```diff
@@ -2300,6 +2300,22 @@
             </properties>
         </profile>
 
+        <profile>
+            <id>flink-1.18</id>
+            <properties>
+                <flink.version>1.18.0</flink.version>
+            </properties>
+            <repositories>
```
Comment (YesOrNo828) [inline pom.xml:2308]: Since the Flink 1.18 is currently in the release candidate stage, the repository will be removed after the stable binary release.

File: .github/workflows/master.yml [hunk 3]
```diff
@@ -183,13 +183,18 @@ jobs:
...
         comment: [ "normal" ]
         include:
           - java: 8
             flink: '1.17'
             flink-archive: '-Dflink.archive.mirror=https://archive.apache.org/dist/flink/flink-1.16.1 -Dflink.archive.name=flink-1.16.1-bin-scala_2.12.tgz'
             comment: 'verify-on-flink-1.16-binary'
+          - java: 8
+            flink: '1.18'
+            flink-archive: '-Dflink.archive.mirror=https://archive.apache.org/dist/flink/flink-1.17.1 -Dflink.archive.name=flink-1.17.1-bin-scala_2.12.tgz'
+            comment: 'verify-on-flink-1.17-binary'
```
Comment (pan3793) [inline .github/workflows/master.yml:197]: Emm... this cross-verification is actually to sure the Flink engine which builds against the default Flink version can work well on all support Flink runtimes.

Specific to this case, it should be: using flink 1.17 for compilation, and verifying that it works well on Flink 1.18 runtime

File: .github/workflows/master.yml [hunk 4]
```diff
@@ -487,7 +492,7 @@ jobs:
       fail-fast: false
       matrix:
         java: [ 8 ]
-        zookeeper: ["3.4", "3.5", "3.6", "3.7" ]
+        zookeeper: [ "3.4", "3.5", "3.6", "3.7" ]
```
Comment (pan3793) [inline .github/workflows/master.yml:495]: ditto.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FlinkEngineUtils.scala [hunk 5]
```diff
@@ -119,7 +119,7 @@ object FlinkEngineUtils extends Logging {
         .build()
         .newInstance(flinkConf, commandLines)
         .asInstanceOf[DefaultContext]
-    } else if (FLINK_RUNTIME_VERSION === "1.17") {
+    } else if (FLINK_RUNTIME_VERSION === "1.17" || FLINK_RUNTIME_VERSION === "1.18") {
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FlinkEngineUtils.scala:122]: how about
```suggestion
    } else if (FLINK_RUNTIME_VERSION >= "1.17") {
```

File: .github/workflows/master.yml (hunk 1 above)
Comment (YesOrNo828) [reply .github/workflows/master.yml]: Due to the 'Action on Save' tool reformat the yml code. I have reverted the cord.

File: .github/workflows/master.yml (hunk 3 above)
Comment (YesOrNo828) [reply .github/workflows/master.yml]: Thanks for your explanation. I've changed the flink version.

File: pom.xml (hunk 2 above)
Comment (pan3793) [reply pom.xml:2308]: it's also required to override `flink.archive.mirror`, it is used to download flink runtime tgz for integration testing

File: .github/workflows/master.yml [hunk 6]
```diff
@@ -183,13 +183,18 @@ jobs:
...
         flink-archive: [ "" ]
         comment: [ "normal" ]
         include:
           - java: 8
             flink: '1.17'
             flink-archive: '-Dflink.archive.mirror=https://archive.apache.org/dist/flink/flink-1.16.1 -Dflink.archive.name=flink-1.16.1-bin-scala_2.12.tgz'
             comment: 'verify-on-flink-1.16-binary'
+          - java: 8
+            flink: '1.17'
+            flink-archive: '-Dflink.archive.mirror=https://archive.apache.org/dist/flink/flink-1.18.0 -Dflink.archive.name=flink-1.18.0-bin-scala_2.12.tgz'
```
Comment (pan3793) [inline .github/workflows/master.yml:196]: the staging repo is required at the RC stage

File: .github/workflows/master.yml [hunk 7]
```diff
@@ -183,13 +183,18 @@ jobs:
...
         flink-archive: [ "" ]
         comment: [ "normal" ]
         include:
           - java: 8
             flink: '1.17'
             flink-archive: '-Dflink.archive.mirror=https://archive.apache.org/dist/flink/flink-1.16.1 -Dflink.archive.name=flink-1.16.1-bin-scala_2.12.tgz'
             comment: 'verify-on-flink-1.16-binary'
+          - java: 8
+            flink: '1.17'
+            flink-archive: '-Dflink.archive.mirror=https://dist.apache.org/repos/dist/dev/flink/flink-1.18.0-rc2 -Dflink.archive.name=flink-1.18.0-bin-scala_2.12.tgz'
```
Comment (pan3793) [inline .github/workflows/master.yml:196]:
```suggestion
            flink-archive: '-Dflink.archive.mirror=https://dist.apache.org/repos/dist/dev/flink/flink-1.18.0-rc3 -Dflink.archive.name=flink-1.18.0-bin-scala_2.12.tgz'
```

**Review by pan3793 — APPROVED** [review]
LGTM, pending CI, thanks for your contribution~

**Comment by pan3793** [comment]
Also cc @link3280 and @SteNicholas

**Comment by link3280** [comment]
LGTM
